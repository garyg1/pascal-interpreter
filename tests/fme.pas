program fme;

// fast modular exponentiation
// adapted from wikipedia
function fme(base, exp, modulus : integer): integer;
var i : integer;
var result : integer = 1;
begin
    while exp > 0 do
    begin
        if (exp mod 2) = 1 then
            result := (result * base) mod modulus;
        exp := exp / 2;
        base := (base * base) mod modulus;
    end;
    fme := result;
end;

begin
    writeln(fme(2, 10, 500));
    writeln(fme(3, 10, 500));
    writeln(fme(3, 1000, 500));
end.
