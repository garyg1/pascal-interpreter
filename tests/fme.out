24
49
1
